packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -2911,6 +2913,40 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     return null;
   }
 
+  TextRange _getParagraphBoundary(TextPosition position) {
+    return TextRange(start: _getParagraphStartOffset(position), end: _getParagraphEndOffset(position));
```

Comment:
> Would it be possible to use the `getLineMetrics` method instead and avoid checking for line break characters? There are other hard line breaks like CR, CLRF, LS and PS.

I'll look into it. The behavior of triple click could select a "paragraph" that wraps and gets rendered across several lines.

There's an `_isWhitespace` helper in this file which checks a hardcoded list of values. I wonder if a helper that considers all of the line breaks in the same way `_isWhitespace` does is the most direct approach.

- TextPainter.computeLineMetrics - processes all the text and returns a bunch of info. In the comments it's said that it could be inefficient but maybe I'm optimizing too early? 
- TextPainter.getLineBoundary - returns the rendered line, so if a paragraph were to occupy 3 lines, only the TextRange of the first line would be returned



> What happens if `position` is surrounded by line breaks? Should it select all the empty lines?

Good question. Just tried this and looks like just the one line (including a line break at the end) is selected.

Reply:
Yeah the reason I suggested using `LineMetrics` instead was that the `Paragraph` implementation already knows where the hard breaks are located, so ideally it should be the source of truth. If the information is stored in a form that's inefficient to look up then feel free to keep the current implementation (in which case eventually we'd like to expose a suitable API on `Paragraph` for this and `_isWhitespace`).

> But it does feel a bit round-about because the TextRange for a "paragraph" doesn't seem to have any dependency on the text layout.

That's right. But if you're doing this specifically for triple tap then it should be fine since hit testing depends on the text layout.

Reply:
Thanks for the explanation. My preference would be to copy `_isWhitespace` for now, but agree this should become part of the `TextPainter` in a future found of refactoring. Wdyt?